Type error in generated code for ivars

Test case:

``` .haskell
import qualified Prelude
import Feldspar
import Feldspar.Compiler

test :: Data Index -> Data Index
test a = share (future (a+1)) $ \a' -> await a' * 2

main = compile test "test" "test" defaultOptions
```

```
> runghc Test.hs
> export FELDSPAR_C=../../../feldspar-compiler/lib/Feldspar/C/
> gcc -O3 -I$FELDSPAR_C -I.. --std=c99 test.c $FELDSPAR_C/ivar.c $FELDSPAR_C/taskpool.c $FELDSPAR_C/feldspar_c99.c -lpthread -lm
test.c: In function ‘task_core1’:
test.c:19:3: warning: passing argument 2 of ‘ivar_put_with_size’ makes pointer from integer without a cast [enabled by default]
In file included from test.h:7:0,
                 from test.c:1:
../../../feldspar-compiler/lib/Feldspar/C/ivar.h:63:6: note: expected ‘void *’ but argument is of type ‘uint32_t’
test.c: In function ‘test’:
test.c:34:3: warning: passing argument 1 of ‘ivar_get_nontask_with_size’ makes pointer from integer without a cast [enabled by default]
In file included from test.h:7:0,
                 from test.c:1:
../../../feldspar-compiler/lib/Feldspar/C/ivar.h:87:6: note: expected ‘void *’ but argument is of type ‘uint32_t’
```

--- a/lib/Feldspar/Compiler/Imperative/FromCore/Future.hs
+++ b/lib/Feldspar/Compiler/Imperative/FromCore/Future.hs
@@ -36,6 +36,8 @@ module Feldspar.Compiler.Imperative.FromCore.Future where
 
 import Language.Syntactic
 
+import Control.Monad.RWS (ask)
+
 import Feldspar.Core.Types (Type,defaultSize)
 import Feldspar.Core.Constructs.Future
 import Feldspar.Core.Interpretation
@@ -52,7 +54,10 @@ instance Compile dom dom => Compile (FUTURE :|| Type) dom
     compileExprSym = compileProgFresh
 
     compileProgSym (C' MkFuture) info (Just loc) (p :* Nil) = do
-        let args = [mkVariable (compileTypeRep t (defaultSize t)) v
+        env <- ask
+        let args = [case lookup v (alias env) of
+                         Nothing -> mkVariable (compileTypeRep t (defaultSize t)) v
+                         Just (VarExpr e) -> e
                    | (v,SomeType t) <- assocs $ infoVars info
                    ] ++ fv loc
         -- Task core:

--- a/lib/Feldspar/Compiler/Imperative/FromCore/Par.hs
+++ b/lib/Feldspar/Compiler/Imperative/FromCore/Par.hs
@@ -34,6 +34,8 @@
 
 module Feldspar.Compiler.Imperative.FromCore.Par where
 
+import Control.Monad.RWS (ask)
+
 import Language.Syntactic
 import Language.Syntactic.Constructs.Monad
 import Language.Syntactic.Constructs.Binding hiding (Variable)
@@ -109,7 +111,10 @@ instance ( Compile dom dom
             tellProg [iVarPut iv var]
 
     compileProgSym ParFork info (Just loc) (p :* Nil) = do
-        let args = [mkVariable (compileTypeRep t (defaultSize t)) v
+        env <- ask
+        let args = [case lookup v (alias env) of
+                         Nothing -> mkVariable (compileTypeRep t (defaultSize t)) v
+                         Just (VarExpr e) -> e
                    | (v,SomeType t) <- assocs $ infoVars info
                    ] ++ fv loc
         -- Task core:

--- a/lib/Feldspar/Compiler/Imperative/Frontend.hs
+++ b/lib/Feldspar/Compiler/Imperative/Frontend.hs
@@ -106,14 +106,14 @@ iVarInit var = call "ivar_init" [ValueParameter var]
 iVarGet :: Expression () -> Expression () -> Program ()
 iVarGet loc ivar 
     | isArray typ   = call "ivar_get_array" [ValueParameter loc, ValueParameter ivar]
-    | otherwise     = call "ivar_get" [TypeParameter typ, ValueParameter loc, ValueParameter ivar]
+    | otherwise     = call "ivar_get" [TypeParameter typ, ValueParameter (AddrOf loc), ValueParameter ivar]
       where
         typ = typeof loc
 
 iVarPut :: Expression () -> Expression () -> Program ()
 iVarPut ivar msg
     | isArray typ   = call "ivar_put_array" [ValueParameter ivar, ValueParameter  msg]
-    | otherwise     = call "ivar_put" [TypeParameter typ, ValueParameter ivar, ValueParameter msg]
+    | otherwise     = call "ivar_put" [TypeParameter typ, ValueParameter ivar, ValueParameter (AddrOf msg)]
       where
         typ = typeof msg
 
@@ -130,10 +130,10 @@ spawn taskName vs = call spawnName allParams
   where
     spawnName = "spawn" ++ show (length vs)
     taskParam = FunParameter taskName
-    typeParams = map (TypeParameter . fixArray . typeof) vs
-      where fixArray (Pointer t@ArrayType{}) = t
-            fixArray t                       = t
-    varParams = map (\v -> ValueParameter $ VarExpr (Variable (typeof v) (vName v))) vs
+    typeParams = map (TypeParameter . typeof) vs
+    varParams = map (ValueParameter . mkv) vs
+      where mkv v = let v' = VarExpr $ Variable (typeof v) (vName v)
+                    in if isPointer $ typeof v then AddrOf v' else v'
     allParams = taskParam : concat (zipWith (\a b -> [a,b]) typeParams varParams)
 
 run :: String -> [Variable ()] -> Program ()
@@ -188,6 +188,10 @@ isNativeArray _             = False
 isIVar :: Type -> Bool
 isIVar IVarType{} = True
 isIVar _              = False
+
+isPointer :: Type -> Bool
+isPointer Pointer{} = True
+isPointer _         = False
 
 dVar :: Declaration () -> Variable ()
 dVar (Declaration v _)    = v

--- a/tests/RegressionTests.hs
+++ b/tests/RegressionTests.hs
@@ -89,6 +89,12 @@ bindToThen y = runMutable $ do
 switcher :: Data Word8 -> Data Bool -> Data Word8
 switcher i = switch (value 0) [(true,i), (false,2)]
 
+ivartest :: Data Index -> Data Index
+ivartest a = share (future (a+1)) $ \a' -> await a' * 2
+
+ivartest2 :: (Data Index, Data Index) -> (Data Index, Data Index)
+ivartest2 a = share (future a) $ \a' -> await a'
+
 tests :: TestTree
 tests = testGroup "RegressionTests"
     [ mkGoldTest example9 "example9" defaultOptions
@@ -102,6 +108,8 @@ tests = testGroup "RegressionTests"
     , mkGoldTest metrics "metrics" defaultOptions
     , mkGoldTest scanlPush "scanlPush" defaultOptions
     , mkGoldTest divConq3 "divConq3" defaultOptions
+    , mkGoldTest ivartest "ivartest" defaultOptions
+    , mkGoldTest ivartest2 "ivartest2" defaultOptions
     , mkBuildTest pairParam "pairParam" defaultOptions
     , mkBuildTest concatV "concatV" defaultOptions
     , mkBuildTest topLevelConsts "topLevelConsts" defaultOptions
@@ -113,6 +121,8 @@ tests = testGroup "RegressionTests"
     , mkBuildTest divConq3 "divConq3" defaultOptions
     , testProperty "bindToThen" (\y -> eval bindToThen y Prelude.== y)
     , mkGoldTest switcher "switcher" defaultOptions
+    , mkBuildTest ivartest "ivartest" defaultOptions
+    , mkBuildTest ivartest2 "ivartest2" defaultOptions
     ]
 
 main :: IO ()
@@ -156,4 +166,3 @@ mkBuildTest fun n opts = do
         cmp _ _ = return Nothing
         upd _ = return ()
     goldenTest n (return "") (liftIO act >> return "") cmp upd
-

--- a/tests/gold/ivartest.c
+++ b/tests/gold/ivartest.c
@@ -1,0 +1,37 @@
+#include "ivartest.h"
+#include "feldspar_c99.h"
+#include "feldspar_array.h"
+#include "feldspar_future.h"
+#include "ivar.h"
+#include "taskpool.h"
+#include <stdint.h>
+#include <string.h>
+#include <math.h>
+#include <stdbool.h>
+#include <complex.h>
+
+
+void task_core1(uint32_t v0, struct ivar v1)
+{
+  uint32_t e0;
+  
+  e0 = (v0 + 1);
+  ivar_put(uint32_t, v1, &e0);
+}
+
+void task1(void * params)
+{
+  run2(task_core1, uint32_t, struct ivar);
+}
+
+void ivartest(uint32_t v0, uint32_t * out)
+{
+  struct ivar v1;
+  uint32_t e2;
+  
+  ivar_init(&v1);
+  spawn2(task1, uint32_t, v0, struct ivar, v1);
+  ivar_get_nontask(uint32_t, &e2, v1);
+  *out = (e2 << 1);
+  ivar_destroy(&v1);
+}

--- a/tests/gold/ivartest.h
+++ b/tests/gold/ivartest.h
@@ -1,0 +1,22 @@
+#ifndef TESTS_IVARTEST_H
+#define TESTS_IVARTEST_H
+
+#include "feldspar_c99.h"
+#include "feldspar_array.h"
+#include "feldspar_future.h"
+#include "ivar.h"
+#include "taskpool.h"
+#include <stdint.h>
+#include <string.h>
+#include <math.h>
+#include <stdbool.h>
+#include <complex.h>
+
+
+void task_core1(uint32_t v0, struct ivar v1);
+
+void task1(void * params);
+
+void ivartest(uint32_t v0, uint32_t * out);
+
+#endif // TESTS_IVARTEST_H

--- a/tests/gold/ivartest2.c
+++ b/tests/gold/ivartest2.c
@@ -1,0 +1,32 @@
+#include "ivartest2.h"
+#include "feldspar_c99.h"
+#include "feldspar_array.h"
+#include "feldspar_future.h"
+#include "ivar.h"
+#include "taskpool.h"
+#include <stdint.h>
+#include <string.h>
+#include <math.h>
+#include <stdbool.h>
+#include <complex.h>
+
+
+void task_core0(struct s_unsignedS32_unsignedS32 * v0, struct ivar v3)
+{
+  ivar_put(struct s_unsignedS32_unsignedS32 *, v3, v0);
+}
+
+void task0(void * params)
+{
+  run2(task_core0, struct s_unsignedS32_unsignedS32 *, struct ivar);
+}
+
+void ivartest2(struct s_unsignedS32_unsignedS32 * v0, struct s_unsignedS32_unsignedS32 * out)
+{
+  struct ivar v3;
+  
+  ivar_init(&v3);
+  spawn2(task0, struct s_unsignedS32_unsignedS32 *, v0, struct ivar, v3);
+  ivar_get_nontask(struct s_unsignedS32_unsignedS32 *, out, v3);
+  ivar_destroy(&v3);
+}

--- a/tests/gold/ivartest2.h
+++ b/tests/gold/ivartest2.h
@@ -1,0 +1,28 @@
+#ifndef TESTS_IVARTEST2_H
+#define TESTS_IVARTEST2_H
+
+#include "feldspar_c99.h"
+#include "feldspar_array.h"
+#include "feldspar_future.h"
+#include "ivar.h"
+#include "taskpool.h"
+#include <stdint.h>
+#include <string.h>
+#include <math.h>
+#include <stdbool.h>
+#include <complex.h>
+
+
+struct s_unsignedS32_unsignedS32
+{
+  uint32_t member1;
+  uint32_t member2;
+};
+
+void task_core0(struct s_unsignedS32_unsignedS32 * v0, struct ivar v3);
+
+void task0(void * params);
+
+void ivartest2(struct s_unsignedS32_unsignedS32 * v0, struct s_unsignedS32_unsignedS32 * out);
+
+#endif // TESTS_IVARTEST2_H

